File: java-sdk/radar-schemas-tools/src/main/java/org/radarbase/schema/tools/ValidatorCommand.java
package org.radarbase.schema.tools;

import static org.radarbase.schema.util.SchemaUtils.applyOrIllegalException;

import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.AbstractMap;
import java.util.Comparator;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Stream;

import kotlin.Pair;
import net.sourceforge.argparse4j.impl.Arguments;
import net.sourceforge.argparse4j.inf.ArgumentParser;
import net.sourceforge.argparse4j.inf.Namespace;
import org.radarbase.schema.SchemaCatalogue;
import org.radarbase.schema.Scope;
import org.radarbase.schema.specification.DataProducer;
import org.radarbase.schema.validation.SchemaValidator;
import org.radarbase.schema.validation.ValidationException;
import org.radarbase.schema.validation.config.ExcludeConfig;
import org.radarbase.schema.validation.rules.SchemaMetadata;

public class ValidatorCommand implements SubCommand {
    @Override
    public String getName() {
        return "validate";
    }

    @Override
    public int execute(Namespace options, CommandLineApp app) {
        try {
            ExcludeConfig config = loadConfig(app.getRoot(), options.getString("config"));

            try {
                System.out.println();
                System.out.println("Validated topics:");
                app.getCatalogue().getSources().stream()
                        .flatMap(s -> s.getData().stream())
                        .flatMap(applyOrIllegalException(d ->
                                d.getTopics(app.getCatalogue().getSchemaCatalogue())))
                        .map(t -> "- "
                                + t.getName()
                                + " [" + t.getKeySchema().getFullName() + ": "
                                + t.getValueSchema().getFullName() + "]")
                        .sorted()
                        .distinct()
                        .forEach(System.out::println);
                System.out.println();
            } catch (Exception ex) {
                System.err.println("Failed to load all topics: " + ex.getMessage());
                return 1;
            }

            String scopeString = options.getString("scope");
            Scope scope = scopeString != null ? Scope.valueOf(scopeString) : null;

            Stream<ValidationException> exceptionStream = Stream.empty();
            SchemaValidator validator = new SchemaValidator(app.getRoot(), config);

            if (options.getBoolean("full")) {
                exceptionStream = validator.analyseFiles(
                        scope,
                        app.getCatalogue().getSchemaCatalogue());
            }
            if (options.getBoolean("from_specification")) {
                exceptionStream = Stream.concat(
                        exceptionStream,
                        validator.analyseSourceCatalogue(scope, app.getCatalogue())).distinct();
            }

            return resolveValidation(exceptionStream, validator,
                    options.getBoolean("verbose"),
                    options.getBoolean("quiet"));
        } catch (IOException e) {
            System.err.println("Failed to load schemas: " + e);
            return 1;
        }
    }

    @Override
    public void addParser(ArgumentParser parser) {
        parser.description("Validate a set of specifications.");
        parser.addArgument("-s", "--scope")
                .help("type of specifications to validate")
                .choices(Scope.values());
        parser.addArgument("-c", "--config")
                .help("configuration file to use");
        parser.addArgument("-v", "--verbose")
                .help("verbose validation message")
                .action(Arguments.storeTrue());
        parser.addArgument("-q", "--quiet")
                .help("only set exit code.")
                .action(Arguments.storeTrue());
        parser.addArgument("-S", "--from-specification")
                .help("validation of all schemas referenced in specifications")
                .action(Arguments.storeTrue());
        parser.addArgument("-f", "--full")
                .help("full validation of contents")
                .action(Arguments.storeTrue());
        SubCommand.addRootArgument(parser);
    }

    private int resolveValidation(Stream<ValidationException> stream,
            SchemaValidator validator,
            boolean verbose,
            boolean quiet) {
        if (quiet) {
            if (stream.count() > 0) {
                return 1;
            }
        } else {
            String result = SchemaValidator.format(stream);

            System.out.println(result);
            if (verbose) {
                System.out.println("Validated schemas:");
                Set<String> names = new TreeSet<>(
                        validator.getValidatedSchemas().keySet());
                for (String name : names) {
                    System.out.println(" - " + name);
                }
                System.out.println();
            }
            if (!result.isEmpty()) {
                return 1;
            }
        }
        return 0;
    }

    private ExcludeConfig loadConfig(Path root, String configSubPath) throws IOException {
        Path configPath;
        if (configSubPath != null) {
            if (configSubPath.charAt(0) == '/') {
                configPath = Paths.get(configSubPath);
            } else {
                configPath = root.resolve(configSubPath);
            }
        } else {
            configPath = null;
        }
        return ExcludeConfig.load(configPath);
    }
}

File: java-sdk/radar-schemas-tools/src/main/java/org/radarbase/schema/tools/ValidatorCommand.kt
package org.radarbase.schema.tools

import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import net.sourceforge.argparse4j.impl.Arguments
import net.sourceforge.argparse4j.inf.ArgumentParser
import net.sourceforge.argparse4j.inf.Namespace
import org.radarbase.schema.Scope
import org.radarbase.schema.tools.SubCommand.Companion.addRootArgument
import org.radarbase.schema.validation.SchemaValidator
import org.radarbase.schema.validation.ValidationException
import org.radarbase.schema.validation.ValidationHelper.COMMONS_PATH
import org.radarbase.schema.validation.toFormattedString
import java.io.IOException
import kotlin.streams.asSequence

class ValidatorCommand : SubCommand {
    override val name: String = "validate"

    override suspend fun execute(options: Namespace, app: CommandLineApp): Int {
        try {
            println()
            println("Validated topics:")
            app.catalogue.sources.asSequence()
                .flatMap { it.data.asSequence() }
                .flatMap { d ->
                    try {
                        d.topics(app.catalogue.schemaCatalogue).asSequence()
                    } catch (ex: Exception) {
                        throw IllegalArgumentException(ex)
                    }
                }
                .distinctBy { it.name }
                .sortedBy { it.name }
                .forEach { topic ->
                    println("- ${topic.name} [${topic.keySchema.fullName}: ${topic.valueSchema.fullName}]")
                }
            println()
        } catch (ex: Exception) {
            System.err.println("Failed to load all topics: " + ex.message)
            return 1
        }

        val scope = options.getString("scope")
            ?.let { Scope.valueOf(it) }

        return try {
            val validator = SchemaValidator(app.root.resolve(COMMONS_PATH), app.config.schemas)

            coroutineScope {
                val fullValidationJob = async {
                    if (options.getBoolean("full")) {
                        if (scope == null) {
                            validator.analyseFiles(app.catalogue.schemaCatalogue)
                        } else {
                            validator.analyseFiles(app.catalogue.schemaCatalogue, scope)
                        }
                    } else {
                        emptyList()
                    }
                }
                val fromSpecJob = async {
                    if (options.getBoolean("from_specification")) {
                        validator.analyseSourceCatalogue(scope, app.catalogue)
                    } else {
                        emptyList()
                    }
                }
                val exceptions = fullValidationJob.await() + fromSpecJob.await()

                resolveValidation(
                    exceptions,
                    validator,
                    options.getBoolean("verbose"),
                    options.getBoolean("quiet"),
                )
            }
        } catch (e: IOException) {
            System.err.println("Failed to load schemas: $e")
            1
        }
    }

    override fun addParser(parser: ArgumentParser) {
        parser.apply {
            description("Validate a set of specifications.")
            addArgument("-s", "--scope")
                .help("type of specifications to validate")
                .choices(Scope.entries)
            addArgument("-v", "--verbose")
                .help("verbose validation message")
                .action(Arguments.storeTrue())
            addArgument("-q", "--quiet")
                .help("only set exit code.")
                .action(Arguments.storeTrue())
            addArgument("-S", "--from-specification")
                .help("validation of all schemas referenced in specifications")
                .action(Arguments.storeTrue())
            addArgument("-f", "--full")
                .help("full validation of contents")
                .action(Arguments.storeTrue())
            addRootArgument()
        }
    }

    private fun resolveValidation(
        stream: List<ValidationException>,
        validator: SchemaValidator,
        verbose: Boolean,
        quiet: Boolean,
    ): Int = when {
        !quiet -> {
            val result = stream.toFormattedString()
            println(result)
            if (verbose) {
                println("Validated schemas:")
                validator.validatedSchemas.keys
                    .sorted()
                    .forEach { println(" - $it") }
                println()
            }
            if (result.isNotEmpty()) 1 else 0
        }
        stream.isNotEmpty() -> 1
        else -> 0
    }
}
